#6726 - Add layers from catalog and filter displayed data

alexander-fedorenko opened the pull request:
## Description
"CATALOG:ADD_LAYERS_FROM_CATALOGS" action and "addLayersFromCatalogsEpic" epic are updated in order to support optional parameter "options". This parameter can be used to overwrite arbitrary options of the resulting layer.

**Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines: https://github.com/geosolutions-it/MapStore2/blob/master/CONTRIBUTING.md
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


**What kind of change does this PR introduce?** (check one with "x", remove the others)
 - [ ] Bugfix
 - [x] Feature
 - [ ] Code style update (formatting, local variables)
 - [ ] Refactoring (no functional changes, no api changes)
 - [ ] Build related changes
 - [ ] CI related changes
 - [ ] Other... Please describe:

<!-- add here the ReadTheDocs link (if needed) -->

## Issue

**What is the current behavior?**
<!-- You can also link to an existing issue here -->
#6726

**What is the new behavior?**
Action supports optional parameter "options". This parameter can be used to overwrite arbitrary options for the created layer, passed object is getting merged with resulting layer object. Following sample can be used to pass CQL_FILTER to the created layer:

URL sample: /#/viewer/openlayers/new?actions=[{"type":"CATALOG:ADD_LAYERS_FROM_CATALOGS","layers":["tiger_roads","tiger_roads","tiger_roads"],"sources":["gs_stable_wms","gs_stable_wms","gs_stable_wms"],"options":[{"params":{"CQL_FILTER":"NAME='5th Ave'"}},{"params":{"CQL_FILTER":"NAME='Brooklyn Brg'"}},{"params":{"CQL_FILTER":"NAME='South End Ave'"}}]},{"type": "ZOOM_TO_EXTENT","extent": [-74.02,40.7,-73.98,40.71],"crs": "EPSG:4326","maxZoom": 15}]

This PR implement following approach:
Added layer will have "cql_filter" property in layer.parameters. It will modify WMS service query and add CQL_FILTER into it. User is able to create more filters for the added layer, all this filters will apply on top of the results obtained with the initial CQL_FILTER passed via action.

## Breaking change
**Does this PR introduce a breaking change?** (check one with "x", remove the other)
 - [ ] Yes, and I documented them in migration notes
 - [x] No

<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications -->

## Other useful information

MV88 commented on web/client/actions/catalog.js:
```diff
@@ -66,12 +66,14 @@ export const SET_FORMAT_OPTIONS = 'CATALOG:SET_FORMAT_OPTIONS';
+ * @param {string[]} filters filters related to each layer
  */
-export function addLayersMapViewerUrl(layers = [], sources = []) {
+export function addLayersMapViewerUrl(layers = [], sources = [], filters = []) {
```
can you change filters to be params?

offtherailz replied:
I said `options` so we can override whatever we want. The final layer have to be simply a deep merge of the 2 JSON

e.g. `options=[{"params":{"CQL_FILTER":"A=B"}}, ...]`

alexander-fedorenko replied:
Thanks! Updated code to use "options" argument.

MV88 commented on web/client/epics/catalog.js:
```diff
@@ -117,7 +117,7 @@ export default (API) => ({
             // maxRecords is 4 (by default), but there can be a possibility that the record desired is not among
             // the results. In that case a more detailed search with full record name can be helpful
-            .switchMap(({ layers, sources, options, startPosition = 1, maxRecords = 4 }) => {
+            .switchMap(({ layers, sources, filters, options, startPosition = 1, maxRecords = 4 }) => {
```
can you change filters to be params?

alexander-fedorenko replied:
Updated code to use existing "options" argument.

MV88 commented on web/client/epics/catalog.js:
```diff
@@ -174,24 +175,35 @@ export default (API) => ({
+                            if (r[1]) {
+                                r[0].params = {
+                                    ...r[0].params,
+                                    CQL_FILTER: r[1]
```
just spread related params object here

in the action the params will be like
/#/viewer/openlayers/new?actions=[{"type":"CATALOG:ADD_LAYERS_FROM_CATALOGS","layers":["tiger_roads","tiger_roads","tiger_roads"],"sources":["gs_stable_wms","gs_stable_wms","gs_stable_wms"],"filters":[{CQL_FILTER: "NAME='5th Ave'"},{ CQL_FILTER: "NAME='Brooklyn Brg'"},{CQL_FILTER:"NAME='South End Ave'"}]}]

alexander-fedorenko replied:
Updated code to use lodash merge to merge default layer options with values passed by "options" object in listed action.

MV88 reviewed (changes requested):
@alexander-fedorenko hi,

me and @offtherailz agreed to:
- use actions from catalog which when I tried it it was working.
- accept the fact that this filter will not be clearable, so no need to have it inside layer filter tool

MV88 commented on web/client/actions/catalog.js:
```diff
@@ -66,12 +66,14 @@ export const SET_FORMAT_OPTIONS = 'CATALOG:SET_FORMAT_OPTIONS';
+ * @param {string[]} options related to each layer. Can be used to overwrite default configuration or parameters
  */
-export function addLayersMapViewerUrl(layers = [], sources = []) {
+export function addLayersMapViewerUrl(layers = [], sources = [], options = []) {
```
please amend or add the test for this action

alexander-fedorenko replied:
Updated existing test

MV88 reviewed (changes requested):
@alexander-fedorenko this point is not being done
- use actions from catalog which when I tried it it was working.

and can you please see if you can add a unit test for the
addLayersFromCatalogsEpic

alexander-fedorenko commented:
@MV88 please re-review.

Following changes are applied:
- Catalog "addLayer" action is in use now.
- Refactored code of "addLayer" thunk, replaced it with the new action + epic.
- Added new tests and updated old ones.

Export of layer that was created by query params will get all data of the layer, including the one that was filtered out via CQL_FILTER parameter.
All additional filters applied to the layer works like if CQL_FILTER from query params and new filter created in TOC works in "AND" conjunction; new filter will be applied on top of the already filtered data.

MV88 reviewed (changes requested):
the filters does not seem to be applied anymore

current
![image](https://user-images.githubusercontent.com/11991428/150559410-140af658-b33c-4eda-8224-49d98eec67f4.png)

expected
![image](https://user-images.githubusercontent.com/11991428/150559369-48337283-feb2-441e-8d5d-798e895dcf5f.png)

alexander-fedorenko commented:
@MV88 I'm sorry, I totally forgot to update PR description after changes. Please check issue body, it has new query with proper structure of "options" parameter.

MV88 reviewed (changes requested):
> /#/viewer/openlayers/new?actions=[{"type":"CATALOG:ADD_LAYERS_FROM_CATALOGS","layers":["tiger_roads","tiger_roads","tiger_roads"],"sources":["gs_stable_wms","gs_stable_wms","gs_stable_wms"],"options":[{"params":{"CQL_FILTER":"NAME='5th Ave'"}},{"params":{"CQL_FILTER":"NAME='Brooklyn Brg'"}},{"params":{"CQL_FILTER":"NAME='South End Ave'"}}]},{"type": "ZOOM_TO_EXTENT","extent": [-74.02,40.7,-73.98,40.71],"crs": "EPSG:4326","maxZoom": 15}]

we were in 2 to forget that part :)

talking about examples... 
can you update the documentation related to query params
this one  https://github.com/geosolutions-it/MapStore2/blob/master/docs/developer-guide/map-query-parameters.md

to include new possibilities?

MV88 commented:
@ElenaGallo please test in DEV

ElenaGallo commented:
Test passed, @alexander-fedorenko please backport to stable branch. Thanks
